distributed/scheduler.py

```diff
@@ -969,40 +1014,56 @@ def __init__(self, name: str):
             self.duration_average = -1
         self.max_exec_time = -1
         self.suspicious = 0
+        self._groups = {}
 
     def add_exec_time(self, duration: float) -> None:
         self.max_exec_time = max(duration, self.max_exec_time)
         if duration > 2 * self.duration_average:
             self.duration_average = -1
 
     def add_duration(self, action: str, start: float, stop: float) -> None:
+        super().add_duration(action, start, stop)
         duration = stop - start
-        self.all_durations[action] += duration
         if action == "compute":
             old = self.duration_average
             if old < 0:
                 self.duration_average = duration
             else:
                 self.duration_average = 0.5 * duration + 0.5 * old
 
+    def transition(self, old: TaskStateState, new: TaskStateState) -> None:
+        super().transition(old, new)
+        self.state_counts[new] += 1
+
+    def add_group(self, tg: TaskGroup) -> None:
+        self._groups[tg] = None
+
+    def remove_group(self, tg: TaskGroup) -> None:
+        # This is important, we need to adjust the stats
+        self._groups.pop(tg)
+        for state, count in tg.states.items():
+            self.states[state] -= count
+        self.duration -= tg.duration
```

Comment:
I guess we've had this problem before but I strongly suspect we'll have to deal with floating point arithmetic precision here.

Historically, we've occasionally encountered negative occupancy issues and maybe this goes back to this diff. 

To avoid this issue entirely we may want to track duration as an integer with `ms` accuracy to avoid this problem. This would require a bit of refactoring and is likely out of scope for this PR.

A quick fix would be to add `self.duration = max(self.duration, 0)` below this

Reply:
I think the duration tracking is only loosely-coupled with other durations in the system, so internally tracking this in ms-resolution and exposing the seconds-variant should be pretty quick to do without far-reaching consequences.

Reply:
https://github.com/dask/distributed/pull/8681/commits/2306ae70907769bac40371119600a6ecf347a436 stores the duration in microseconds internally. (I think that's the smallest precision we get, mulliseconds is too coarse-grained and causes test failures.)